drivers:flash:atmel SAM0 fix flash_write so it can handle smaller writes

photonthunder opened the pull request:
flash_sam0_write currently only allows page writes of FLASH_PAGE_SIZE.  Following advice from @Laczen I used the flash_sam_write as guide for how to modify flash_sam0_write.  MCUBOOT will not work with the sam0 processors without this modification since it tries to write 8 or 16 bytes of header information.

Tested on the SAMD51 by doing a MCUBOOT SWAP/MOVE and update was completed successfully.

photonthunder committed:
drivers: flash: atmel SAM0 fix flash_write to handle smaller length.

Fix flash_sam0_write to handle byte lengths smaller than FLASH_PAGE_SIZE

Signed-off-by: Daniel Evans <photonthunder@gmail.com>

Laczen commented:
@photonthunder, thanks for the PR. Looks like a simple and efficient fix. Have you verified that the test in tests/drivers/flash/common is succesfull for the atmel sam0 ?

pdgendt commented on drivers/flash/flash_sam0.c:
```diff
@@ -163,10 +165,11 @@ static int flash_sam0_write_page(const struct device *dev, off_t offset,
 	/* Ensure writes happen 32 bits at a time. */
-	for (; src != end; src++, dst++) {
-		*dst = UNALIGNED_GET((uint32_t *)src);
+	for (; len > 0; len -= sizeof(*src)) {
```
Minor concern: add a check or at least an assert that `len % sizeof(*src) == 0U` to as a sanity check to prevent an infinite loop.

attie-argentum replied:
Indeed, what's wrong with the original? (this change will fail and `len` will wrap to "_very positive_" if its not a multiple of 4...)

attie-argentum replied:
It's handled in [`flash_sam0_write()`](https://github.com/zephyrproject-rtos/zephyr/blob/19ef21d91cd11a905e2460952556a1da7811877d/drivers/flash/flash_sam0.c#L297) using a multiple of `FLASH_WRITE_BLK_SZ`, but this pushes the relevant checking / info out-of-file and into the device trees...

photonthunder replied:
The idea was to mirror as much as possible the method implemented in flash_sam.c.  The same risk would exist in that code as well...

attie-argentum reviewed (changes requested):
Please also resolve the call to [`flash_sam0_write_page()` from `flash_sam0_commit()`](https://github.com/zephyrproject-rtos/zephyr/blob/19ef21d91cd11a905e2460952556a1da7811877d/drivers/flash/flash_sam0.c#L217-L219) - the `len` parameter needs to be added.

... that said, would MCUBOOT be happy with (or even prefer) `CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES`, instead of making a "_write page_" function deal with non-pages?

Laczen commented:
> Please also resolve the call to [`flash_sam0_write_page()` from `flash_sam0_commit()`](https://github.com/zephyrproject-rtos/zephyr/blob/19ef21d91cd11a905e2460952556a1da7811877d/drivers/flash/flash_sam0.c#L217-L219) - the `len` parameter needs to be added.
> 
> ... that said, would MCUBOOT be happy with (or even prefer) `CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES`, instead of making a "_write page_" function deal with non-pages?

That would be possible, but it would leave the problem of the write-block-size being defined as 4 while the driver considers it to be 256. And in reality the write-block-size does not seem to be 256 at all.

attie-argentum commented:
>> ... that said, would MCUBOOT be happy with (or even prefer) CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES, instead of making a "write page" function deal with non-pages?

> That would be possible, but it would leave the problem of the write-block-size being defined as 4 while the driver considers it to be 256. And in reality the write-block-size does not seem to be 256 at all.

[`CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES` presents a write block size of 1 byte](https://github.com/zephyrproject-rtos/zephyr/blob/bbec614b78b3a083fc7a559ae3ac00a5af24973d/drivers/flash/flash_sam0.c#L68), unless I'm misunderstanding?

These parts have row-erase, and page-write granularity - for the SAML21 family, pages are 64 bytes, rows are 4x pages (i.e: 256 bytes)... To behave correctly, do we not need to declare the larger of the two? ([there isn't a distinction between erase/write in `struct flash_parameters`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/flash.h#L57-L60))

photonthunder commented:
> @photonthunder, thanks for the PR. Looks like a simple and efficient fix. Have you verified that the test in tests/drivers/flash/common is succesfull for the atmel sam0 ?

uart:~$ *** Booting Zephyr OS build zephyr-v3.4.0-2-g0cbffa53a5b3 ***
Running TESTSUITE flash_driver
===================================================================
START - test_read_unaligned_address
 PASS - test_read_unaligned_address in 0.049 seconds
===================================================================
TESTSUITE flash_driver succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [flash_driver]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.049 seconds
 - PASS - [flash_driver.test_read_unaligned_address] duration = 0.049 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL

photonthunder commented:
@Laczen, ran test zephyr/drivers/flash/common on my SAMD51 board and it passed:
SUITE PASS - 100.00% [flash_driver]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.049 seconds
 - PASS - [flash_driver.test_read_unaligned_address] duration = 0.049 seconds

photonthunder commented:
> Please also resolve the call to [`flash_sam0_write_page()` from `flash_sam0_commit()`](https://github.com/zephyrproject-rtos/zephyr/blob/19ef21d91cd11a905e2460952556a1da7811877d/drivers/flash/flash_sam0.c#L217-L219) - the `len` parameter needs to be added.
> 
> ... that said, would MCUBOOT be happy with (or even prefer) `CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES`, instead of making a "_write page_" function deal with non-pages?

Fixed and did compile with CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES to make sure...

Laczen commented:
> > > ... that said, would MCUBOOT be happy with (or even prefer) CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES, instead of making a "write page" function deal with non-pages?
> 
> > That would be possible, but it would leave the problem of the write-block-size being defined as 4 while the driver considers it to be 256. And in reality the write-block-size does not seem to be 256 at all.
> 
> [`CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES` presents a write block size of 1 byte](https://github.com/zephyrproject-rtos/zephyr/blob/bbec614b78b3a083fc7a559ae3ac00a5af24973d/drivers/flash/flash_sam0.c#L68), unless I'm misunderstanding?
> 
> These parts have row-erase, and page-write granularity - for the SAML21 family, pages are 64 bytes, rows are 4x pages (i.e: 256 bytes)... To behave correctly, do we not need to declare the larger of the two? ([there isn't a distinction between erase/write in `struct flash_parameters`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/flash.h#L57-L60))

@attie-argentum, to behave correctly the erase-block-size should be the smallest size that allows erase, the write-block-size should be the smallest size that allows a write. As far as I understand the specifications for the atmel devices they allow writing blocks of 4 bytes but it is not allowed to do a write that is larger than a "page", so writes longer than a page should be split up in several "page writes" (the routine "write_page" should be understood as "write_in_page"). Zephyr has no equivalent parameter for the page-size (also not on spi-nor devices), this page-size would have the meaning of the largest block size that can be written in one "step", typically limited by a buffer size on spi-nor devices.

Laczen commented on drivers/flash/flash_sam0.c:
```diff
@@ -67,7 +70,7 @@ static const struct flash_parameters flash_sam0_parameters = {
 	.write_block_size = 1,
 #else
-	.write_block_size = DT_PROP(DT_INST(0, soc_nv_flash), write_block_size),
+	.write_block_size = FLASH_WRITE_BLK_SZ,
```
You could add `BUILD_ASSERT((FLASH_WRITE_BLK_SZ % sizeof(uint32_t)) == 0, "unsupported write-block-size");`

photonthunder replied:
Added it right before flash_sam0_write...

nandojve commented:
Hi @attie-argentum ,

I think the improvement is relevant and I marked for v3.5. I`ll start to look after all your concerns be solved.

de-nordic commented:
@attie-argentum Can you take a look again?

attie-argentum reviewed (commented):
A couple of comments from me, but nothing worth blocking.

I'm also still a uneasy that `flash_sam0_write_page()` can be called in a way that may cause an exception without catching it in software first - checking `offset` and `len` inside that function might be very sensible (possibly there instead of in `flash_sam0_write()`, but please consider byte emulation too)... It does seem to be well-defined as far as external interfaces go, but future modifications within this file may miss this. Please add a comment at the very least.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/36298367/2e0f8188-fa13-4ca6-975c-6c68a3061da8)

photonthunder commented:
@attie-argentum added your changes.  I added the note as well.

attie-argentum reviewed (dismissed):
Thanks!

nandojve commented on drivers/flash/flash_sam0.c:
```diff
@@ -163,10 +173,11 @@ static int flash_sam0_write_page(const struct device *dev, off_t offset,
+		*dst++ = UNALIGNED_GET((uint32_t *)src++);
 	}
 
+
```
Drop space line 180

nandojve commented on drivers/flash/flash_sam0.c:
```diff
@@ -222,6 +233,8 @@ static int flash_sam0_commit(const struct device *dev, off_t base)
 	return 0;
 }
 
+
+
```
Too many space lines at 236, 237

nandojve reviewed (changes requested):
Please add this to [zephyr](https://github.com/zephyrproject-rtos/zephyr/tree/f3f183a66cee292ced84866a0b3023f08b44a688)/[tests](https://github.com/zephyrproject-rtos/zephyr/tree/f3f183a66cee292ced84866a0b3023f08b44a688/tests)/[drivers](https://github.com/zephyrproject-rtos/zephyr/tree/f3f183a66cee292ced84866a0b3023f08b44a688/tests/drivers)/[flash](https://github.com/zephyrproject-rtos/zephyr/tree/f3f183a66cee292ced84866a0b3023f08b44a688/tests/drivers/flash)

photonthunder commented:
@nandojve, fixed the spaces.  Not sure what you mean by adding to zephyr/tests/drivers/flash?

Here is the output of the test:

*** Booting Zephyr OS build zephyr-v3.4.0-4-gfe75d1521a69 ***
Running TESTSUITE flash_driver
===================================================================
START - test_read_unaligned_address
 PASS - test_read_unaligned_address in 0.049 seconds
===================================================================
TESTSUITE flash_driver succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [flash_driver]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.049 seconds
 - PASS - [flash_driver.test_read_unaligned_address] duration = 0.049 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL

nandojve commented:
> @nandojve, fixed the spaces. Not sure what you mean by adding to zephyr/tests/drivers/flash?

I would make sure it is running on CI for build.

photonthunder commented:
> >  would make sure it is running on CI for build.

@nandojve, I am assuming you mean running something like the following:

west twister -p ${MY_BOARD} -T tests/drivers/flash

`
INFO    - Total complete:   10/  10  100%  skipped:    9, failed:    0, error:    0
INFO    - 10 test scenarios (10 test instances) selected, 9 configurations skipped (8 by static filter, 1 at runtime).
INFO    - 1 of 10 test configurations passed (100.00%), 0 failed, 0 errored, 9 skipped with 0 warnings in 16.09 seconds
INFO    - In total 1 test cases were executed, 10 skipped on 1 out of total 576 platforms (0.17%)
INFO    - 0 test configurations executed on platforms, 1 test configurations were only built.
`

nandojve commented:
> > > would make sure it is running on CI for build.
> 
> @nandojve, I am assuming you mean running something like the following:
> 
> west twister -p ${MY_BOARD} -T tests/drivers/flash
> 
> INFO - Total complete: 10/ 10 100% skipped: 9, failed: 0, error: 0 INFO - 10 test scenarios (10 test instances) selected, 9 configurations skipped (8 by static filter, 1 at runtime). INFO - 1 of 10 test configurations passed (100.00%), 0 failed, 0 errored, 9 skipped with 0 warnings in 16.09 seconds INFO - In total 1 test cases were executed, 10 skipped on 1 out of total 576 platforms (0.17%) INFO - 0 test configurations executed on platforms, 1 test configurations were only built.

No.

One thing is running twister manually the other is configure to run on the Zephyr infrastructure.

CI is configured by the yaml files like [testcase.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/flash/common/testcase.yaml)
The intention is configure the board/testcase to make sure when `flash` or `drivers` tags are triggered this driver is built.

photonthunder commented:
> CI is configured by the yaml files like [testcase.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/flash/common/testcase.yaml) The intention is configure the board/testcase to make sure when `flash` or `drivers` tags are triggered this driver is built.

@nandojve, I had looked at that but it has no reference to any sam or sam0 boards and for me it isn't intuitive what I would add to make it work.  The boards I work on are all proprietary and will never be added as official Zephyr boards.

Tell me what to add and I will be happy to add it...

de-nordic commented on drivers/flash/flash_sam0.c:
```diff
@@ -163,8 +173,8 @@ static int flash_sam0_write_page(const struct device *dev, off_t offset,
-	for (; src != end; src++, dst++) {
-		*dst = UNALIGNED_GET((uint32_t *)src);
+	for (; len > 0; len -= sizeof(*src)) {
+		*dst++ = UNALIGNED_GET((uint32_t *)src++);
```
`UNALIGNED_GET` is a macro, don't do `++` with a macro parameter.

Laczen replied:
@de-nordic, very nice catch !

photonthunder replied:
@de-nordic, what was your thought with end?  It is possible as with the byte implementation, as @attie-argentum mentioned previously, for the length to not be intervals of sizeof(uint32_t)?

attie-argentum replied:
If `len` is not a multiple of 4, then you'll already be losing the last portion of the write... in fact, you cannot write 3 bytes (only 1, 2 or 4 bytes IIRC). This is guarded against by A) the comment on `flash_sam0_write_page()`, and B) the use of `ROW_SIZE` and `(len % FLASH_WRITE_BLK_SZ) != 0` (depending on the route in).

The issue here is that [`UNALIGNED_GET()` is a macro](https://github.com/zephyrproject-rtos/zephyr/blob/9ca543d7ae2a8bed8b4ecfd3379b3e222b81b8ae/include/zephyr/toolchain/gcc.h#L125-L132), and thus may evaluate the `src++` more than once(!)

I'd prefer to keep the original, with a patch on the `end` variable:

```c
const uint32_t *src = data;
const uint32_t *end = src + (len / sizeof(*src));

// ...

for (; src != end; src++, dst++) {
    *dst = UNALIGNED_GET((uint32_t *)src);
}
```

photonthunder replied:
@attie-argentum, I am following now.  Thanks for the explanation.

attie-argentum replied:
Great! no worries 😊

photonthunder commented:
So with all changes applied I know get a verify error (flash_sam0: verify error at offset) when you write anything less than FLASH_PAGE_SIZE.  The write is successful so there must be something subtle with the following that I don't see:

	if (memcmp(data, FLASH_MEM(offset), len) != 0) {
		LOG_ERR("verify error at offset 0x%lx", (long)offset);
		return -EIO;
	}

attie-argentum commented:
> Due to the fact that len was stepped down to 0 this verification was never run.

Well spotted - it would indeed have run with `len=0` in your previous patchset.

> I think it should probably just be removed all together since verification is probably best done elsewhere anyway?

I can't immediately see why it wouldn't be working, but I do not think it should be removed.

> The write is successful so there must be something subtle with the following that I don't see.

Have you verified the flash contents elsewhere in the same way and it's succecssful? (or dumped it?)

photonthunder commented:
@Laczen, thanks for the idea.  I moved it but it gave the same error.

I ran a simple experiment where I took the 16 or 8 byte values that are being written as part of MCUBOOT test/confirm and shifted them.  If I did offset -= 512, then the validation step would pass.  Same for 256, 128, and 64.  When I get to 32 (offset -= 32) then it no longer returns the correct value, just all FF.  This is the same thing it does with no shift, or in other words using the current code.  I have confirmed via read of the flash that the data does get written every time, it just doesn't read correctly with that particular verification method.  I have tried a few different ways but nothing has worked so far.

Laczen commented:
@photonthunder, shouldn't `data` be casted to a `uint32_t *` before calling memcmp ?

attie-argentum commented:
> > Have you verified the flash contents elsewhere in the same way and it's succecssful? (or dumped it?)
>
> @attie-argentum, the verification when len is FLASH_PAGE_SIZE always works, but any other length (8 or 16 bytes) it will fail every time even though the values are valid.

I was trying to hone in on your claim of "_the values are valid_" - if the `memcmp()` is failing, this would appear to be an incorrect statement... Have you done a `memcmp()` elsewhere, or dumped the region of flash to verify it in some way outside of this function? (e.g: by disabling this `memcmp()`, comparing [in the test](https://github.com/zephyrproject-rtos/zephyr/blob/67e3c21/tests/drivers/flash/common/src/main.c#L131-L133))

Things appear to behave for me, running `.../tests/drivers/flash/common/` passes, and does so even if I apply the patch below to restrict it to 4 byte writes.

```patch
diff --git tests/drivers/flash/common/src/main.c tests/drivers/flash/common/src/main.c
index 4e73617..2b039e4 100644
--- tests/drivers/flash/common/src/main.c
+++ tests/drivers/flash/common/src/main.c
@@ -45,7 +45,7 @@
 #error "Unsupported configuraiton"
 #endif

-#define EXPECTED_SIZE  512
+#define EXPECTED_SIZE  4
 #define CANARY         0xff

 static const struct device *const flash_dev = TEST_AREA_DEVICE;
@@ -114,6 +114,7 @@ ZTEST(flash_driver, test_read_unaligned_address)
                         expected, EXPECTED_SIZE);
        zassert_equal(rc, 0, "Cannot write to flash");

+#if 0
        /* read buffer length*/
        for (off_t len = 0; len < 25; len++) {
                /* address offset */
@@ -143,6 +144,7 @@ ZTEST(flash_driver, test_read_unaligned_address)
                        }
                }
        }
+#endif
 }

 ZTEST_SUITE(flash_driver, NULL, flash_driver_setup, NULL, NULL, NULL);
```

attie-argentum commented:
... I find myself circling back to suggesting you use `CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES` with MCUBOOT. This gives the illusion of byte-sized pages (via RMW), but I acknowledge it's far from ideal if MCUBOOT will erase a page and then perform lots of small writes as a second step.

> to behave correctly the erase-block-size should be the smallest size that allows erase, the write-block-size should be the smallest size that allows a write

@Laczen - given that `struct flash_parameters` only has a `.write_block_size`, and no erase size, what's the best way to communicate these two different sizes to users? (this patch still declares the larger of the two as the value of `flash_parameters.write_block_size`)

photonthunder commented:
@attie-argentum, I hacked the flash test code so I could put in specific addresses and I can confirm that it is always successful regardless of location and/or value.

@Laczen, I see the verification failure whenever I do a mcuboot boot_request_upgrade.  If I take out the return -EIO; line so it just shows the error but doesn't return, then boot_request_upgrade returns successfully, further confirming that the data is being written correctly.

In my case with my SAMD51 board boot_request_upgrade will write 16 bytes of flash at the last of slot 1 and then write 8 bytes before it.  What I see on occasion is I get a verification error only on the first 16 byte write and then the second 8 byte write will sometimes not get a verification error, but it is random. 

Is FLASH_MAP or some other feature causing a delayed write, thus making a quick verification read erroneous?

photonthunder commented:
Further developing on the thought that it needs more time before performing a read I placed the verification at the end of flash_sam0_write with a 1 MS sleep:
static int flash_sam0_write(const struct device *dev, off_t offset,
			    const void *data, size_t len)
{
	size_t verify_len = len;
	size_t verify_offset = offset;

...

	flash_sam0_sem_give(dev);
	k_sleep(K_MSEC(1));

	if (memcmp(data, ((uint8_t *)CONFIG_FLASH_BASE_ADDRESS + verify_offset), verify_len) != 0) {
		LOG_ERR("verify error at offset 0x%lx", (long)verify_offset);
		//return -EIO;
	}

Then I no longer get any verification errors.  If I remove the sleep I still get the verification error.

@attie-argentum, @Laczen I am still of the thought that we just take it out?  To me, verification is something you do at a higher level...

Laczen commented:
@photonthunder, could there be an error in the ready status evaluation?

photonthunder commented:
@Laczen, you inspired me to pull out MPLAB and look at the flash write drivers.  What I noticed is the following:

#define   NVMCTRL_CTRLA_WMODE_MAN_Val     _U_(0x0)   /**< \brief (NVMCTRL_CTRLA) Manual Write */
#define   NVMCTRL_CTRLA_WMODE_ADW_Val     _U_(0x1)   /**< \brief (NVMCTRL_CTRLA) Automatic Double Word Write */
#define   NVMCTRL_CTRLA_WMODE_AQW_Val     _U_(0x2)   /**< \brief (NVMCTRL_CTRLA) Automatic Quad Word */
#define   NVMCTRL_CTRLA_WMODE_AP_Val      _U_(0x3)   /**< \brief (NVMCTRL_CTRLA) Automatic Page Write */

bool NVMCTRL_DoubleWordWrite(const uint32_t *data, const uint32_t address)
{
    uint8_t i = 0;
    bool wr_status = false;
    uint32_t * paddress = (uint32_t *)address;
    uint32_t wr_mode = (NVMCTRL_REGS->NVMCTRL_CTRLA & NVMCTRL_CTRLA_WMODE_Msk); 

    /* Clear global error flag */
    nvm_error = 0;

    /* If the address is not a double word address, return error */
    if((address & 0x07) != 0)
    {
        wr_status = false;
    }
    else
    {
        /* Configure Double Word Write */
        NVMCTRL_SetWriteMode(NVMCTRL_WMODE_ADW);

        /* Writing 32-bit data into the given address.  Writes to the page buffer must be 32 bits. */
        for (i = 0; i <= 1; i++)
        {
            *paddress++ = data[i];
        }
        /* Restore the write mode */
        NVMCTRL_SetWriteMode((NVMCTRL_WRITEMODE)wr_mode);
        wr_status = true;
    }
    return wr_status;
}

I "assume" it is just timing out after a certain period when you don't write the full page.  I will work on a write version for 8 byte writes or double word in this case....

Laczen commented:
@photonthunder, in MPLAB the word seems to be 16bit, double word 32bit (4 bytes) and quad word would then be 64bit (8bytes). This could finally explain the strange behavior.

photonthunder commented:
@Laczen, implemented the double word write and the problem is still identical.  I thought for sure that would fix it...

Laczen commented:
> @Laczen, implemented the double word write and the problem is still identical. I thought for sure that would fix it...

DAMNED, I was hoping the same. While going trough the code I have seen that during init the write mode is set to manual based upon the definition of `NVMCTRL_CTRLB_MANW`, IIRC you are using a samd51, but `NVMCTRL_CTRLB_MANW` is not defined in the hal for this device. Could it be that support for samd51 has not been tested properly ?

nandojve commented:
> DAMNED, I was hoping the same. While going trough the code I have seen that during init the write mode is set to manual based upon the definition of NVMCTRL_CTRLB_MANW, IIRC you are using a samd51, but NVMCTRL_CTRLB_MANW is not defined in the hal for this device. Could it be that support for samd51 has not been tested properly ?

MCHP is well know to miss this kind of info on the headers.

> Could it be that support for samd51 has not been tested properly ?

I don't believe on that because there are bootloaders available for SAMD5x.

I'm not looking close this issue at moment so if the info is irrelevant, just ignore it but I remember that SAMV7 requires 16 bytes alignment to work. I remember that SAMV7 take a very long time to have support to MCUboot because of that. This means that when you sign the binary you must pass the correct value 16 instead use default ones. It won't be a surprise to me if for SAMD5x requires 16 bytes alignment. So, if you didn't checked yet I would recommend increase to 16 bytes and make a try.

de-nordic commented:
As far as I can tell the MCUboot may support alignment higher than 32, it has not been tested (that is why it is guarded against setting anything larger than 32).
The "magic" is able to store alignment size and code is using BOOT_MAX_ALIGN where work with alignment is needed.
Maybe it is worth checking first.

de-nordic commented on drivers/flash/flash_sam0.c:
```diff
@@ -301,12 +311,21 @@ static int flash_sam0_write(const struct device *dev, off_t offset,
+		while (len > 0) {
+			size_t eop_len, write_len;
+
+			/* Maximum size without crossing a page */
+			eop_len = FLASH_PAGE_SIZE - (offset % FLASH_PAGE_SIZE);
```
This part is important only for the first loop run so it can be as well executed outside of loop, before entering it for the first time; when moved, the `eop_len` can just become pre-initialized `write_len`, that will be corrected, when needed, by line 319.

photonthunder replied:
@de-nordic, say you were writing FLASH_PAGE_SIZE + 8 bytes.  If that is the case, and since we increment offset, you would want to calculate eop_len on every loop?  If you change eop_len to FLASH_PAGE_SIZE then what you stated would work (which may be what you are suggesting)?

de-nordic replied:
No. As far as I understand you predict that somebody will try to write, let say, `len > FLASH_PAGE_SIZE`, which means that you have full page + maybe prefix of bytes and maybe suffix of bytes. Something like 
`start_bytes + N * FLASH_PAGE_SIZE + end_bytes`, where `0<= start_bytes < FLASH_PAGE_SIZE`, `0 <= end_bytes  < FLASH_PAGE_SIZE` and N is [0, whatever].
You can not cross page boundary so you have to write `start_bytes` first, then you can have  N page aligned chunks  and then write `end_bytes`.
So your first `write_len` would be `start_bytes`, and you can calculate it before entering the loop; then you are aligned and can write `write_len == FLASH_PAGE_SIZE` N times and finally write `end_bytes`.

So within loop, after you update `len -= write_len`, as you do now, you can just update `write_len = MIN(FLASH_PAGE_SIZE, len)`.
With the simple change you remove some of calculations out of the loop.
But it is Friday and I may have got that wrong.

photonthunder replied:
I hear you about Friday.  I think I get the logic now.  I pushed an updated version of the code...

attie-argentum commented:
The SAMD51 datasheet has more/different words on the minimum write topic... 16-bit or 8-bit reads are not allowed (where they are on SAML21)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/36298367/6ec7a327-dd5b-443d-ad98-45719cbf7b33)


The following is also important for this work... i.e: minimum "safe" write is 128-bit, as @nandojve suggested, but the order that data is written within the 8-byte region doesn't matter.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/36298367/ef3167c6-7c44-4445-acd1-44f4b59e4dfb)


@photonthunder - are your writes smaller than 128-bit / 8-bytes?


---

> Something that worries me is that the flash_sam (not flash_sam0) driver is not doing any verification at all, to be honest most drivers do not check this.

If this is something that others don't verify, and becomes correct after a small delay... do we take the "risk" and remove verification as discussed? It's been there since day one without issue...

> BTW: the byte write emulation is mostly done using a read, modify, erase, write and this: a. Does to many erases, b. Could lead to data loss when the power is interrupted during a erase. This method should be avoided if possible.

Now that I've remembered what MCUBOOT is doing, I agree... ignore that suggestion.

> could there be an error in the ready status evaluation?

This looks okay to me, but please feel free to verify - [`flash_sam0_wait_ready()`](https://github.com/zephyrproject-rtos/zephyr/blob/d323c19406d76dabb5ace3ff3392d8923010cc7f/drivers/flash/flash_sam0.c#L113) / [`flash_sam0_check_status()`](https://github.com/zephyrproject-rtos/zephyr/blob/d323c19406d76dabb5ace3ff3392d8923010cc7f/drivers/flash/flash_sam0.c#L124)

> I "assume" it is just timing out after a certain period when you don't write the full page. I will work on a write version for 8 byte writes or double word in this case....

The driver sets `MANW` (if available), and uses `PBC` / `WP` to clear the page buffer and then write the page respectively... on SAMD51, `CTRLB.bit.MANW` isn't available, but it uses `CTRLA.bit.WMODE` instead (should be `0x0` / Manual by default, but isn't set by the driver...)

Perhaps add this to `flash_sam0_init()` to make sure:

```patch
 #ifdef NVMCTRL_CTRLB_MANW
     /* Require an explicit write command */
     NVMCTRL->CTRLB.bit.MANW = 1;
+#elif NVMCTRL_CTRLA_WMODE
+    /* Set manual write mode */
+    NVMCTRL->CTRLA.bit.WMODE = NVMCTRL_CTRLA_WMODE_MAN_Val;
 #endif
```

Laczen commented:
> As far as I can tell the MCUboot may support alignment higher than 32, it has not been tested (that is why it is guarded against setting anything larger than 32). The "magic" is able to store alignment size and code is using BOOT_MAX_ALIGN where work with alignment is needed. Maybe it is worth checking first.

In theory this is true, but... while swapping the status is written to flash for each sector processed and this status takes up a write-block-size (or 4 bytes if this is larger). For a device with many small sectors this will ruin a lot of flash space or it might even be impossible to do a write.

The problem is also not mcuboot, @photonthunder has gotten this to work, the problem is the flash driver: the verification errors out (but correct data is written). I am pleased by the approach of @photonthunder: the problem is not ignored (although it would be very easy to remove the verification (no other driver is doing this) and just to hope everything works). That said, if this problem isn't solved in a reasonable time I would propose just to remove the verification.

Laczen commented:
@attie-argentum, agreed that it should default to manual mode. The 128bit is 16byte not 8byte. Looking at the datasheet again the device has an ecc over 64bit, so for a correct read to be performed at least 64bit should be written.

attie-argentum commented:
> The 128bit is 16byte not 8byte

Ahem, yes... thanks 🤦‍♀️

photonthunder commented:
@nandojve, @de-nordic I did try 16 and 32 for the write-block-size and the same issue.  Writes fine but the read verification fails.  Interesting thing is that the read verification failure doesn't occur for every write, just select ones.

@attie-argentum, your quote from the manual stating that partial writes can be done with manual mode confirms why it works for none page writes.  I did add the default as you suggested, but it made no difference as you suspected...

Are we ok with just removing the verification?  I am also ok leaving it as a warning if that is preferred.  A write has never failed, just the immediate read verification.  I am going to reach out to Microchip and see if they have any insight but I don't have any other things to try?

de-nordic commented:
> Writes fine but the read verification fails. Interesting thing is that the read verification failure doesn't occur for every write, just select ones.
> 
 
Just from curiosity: may it happen that there have been double write to page that caused the issue? As I understand the page buffer is filled with erase value (1s) every time write from buffer is committed, before it gets modified by write; this means that the page buffer may differ from the actual page it will be written to.

photonthunder commented:
> Just from curiosity: may it happen that there have been double write to page that caused the issue? As I understand the page buffer is filled with erase value (1s) every time write from buffer is committed, before it gets modified by write; this means that the page buffer may differ from the actual page it will be written to.

@de-nordic, good thought.  Here is the output when I try request upgrade.  This is with write-block-size = <8>.  
The Flash offset line will run every time flash_sam0_write_page is called.  In this case only the 16 byte write verification fails, the 8 byte does not fail.  This appears to be pseudo random where a change in address or length and I get different results.

[00:00:00.078,000] <wrn> flash_sam0: Flash offset = 0xFFFF0, len = 16
[00:00:00.078,000] <wrn> flash_sam0: verify error at offset 0xffff0
[00:00:00.078,000] <wrn> flash_sam0: Flash offset = 0xFFFD8, len = 8

photonthunder commented:
Pushed a version with all the latest code tweaks.  I have changed the verification to a warning and commented out the return error.  Not expecting it to be a final but thought it would spike discussion.

photonthunder commented:
I was concerned that there was something specific to the board design I was using that may be causing this issue.  I just tested on a new design with a SAME53 instead of a SAMD53 and I still see the same issue.

nandojve commented:
Hi @photonthunder ,

Would you mind to test the driver disabling cache controller just to make sure it is no part of the problem.

photonthunder commented:
> Would you mind to test the driver disabling cache controller just to make sure it is no part of the problem.

Good thought.  Just to make sure we are on the same page.  You are referring to the CMCC cache controller referenced in section 11 of the data sheet: 

11.6.2 Initialization and Normal Operation
On reset, the cache controller data entries are all invalidated, and the cache is disabled. The cache is transparent to processor operations. The cache controller is activated through the use of its configuration registers. The configuration interface is memory mapped in the APB bus.
Use the following sequence to enable the cache controller:
• Verify that the CMCC is disabled, reading the value of the SR.CSTS.
• Enable the CMCC by writing '1' in CTRL.CEN. The MODULE is disabled by writing a '0' in CTRL.CEN.

Nothing has been enabled on purpose but I will double check...

nandojve commented:
I'm talking about this, to be specific:

https://github.com/zephyrproject-rtos/zephyr/blob/fce06403364b1330a8dd2b72a148a6c0cc347e1c/soc/arm/atmel_sam0/common/soc_samd5x.c#L119-L120

photonthunder commented:
> I'm talking about this, to be specific:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/fce06403364b1330a8dd2b72a148a6c0cc347e1c/soc/arm/atmel_sam0/common/soc_samd5x.c#L119-L120

I have no idea how I missed that, will do some testing...

photonthunder commented:
That was it (still can't believe I missed that).  So I no longer get the error if cache is off.

I am still of the mindset this verification step gets removed since no other flash_writes do a read.  Thoughts?

nandojve commented:
Hi @photonthunder ,

Thank you for the news! I expect that now the verification will pass. It is very important to keep it because cache probably will be enabled in the future and the verification is part of the game to make it work properly.

Just to inform, I already sent a patch to temporarily disable the cache at #61005.

photonthunder commented:
> Just to inform, I already sent a patch to temporarily disable the cache at #61005.

Thanks @nandojve, I pull the verification back the way it was...

Laczen commented:
@photonthunder, that is really good news.

photonthunder commented:
Formatting Fixed...

nandojve commented:
The #61005 got merged.

I would like know what is the test that we can verify on real HW that everything is OK.
Do I need to test using MCUboot too or there is a test in Zephyr that can simulate the required small write blocks?

I would like to have this driver on CI for at least one board with SAME5x, SAMD20, SAMR21 (SAMD21) and SAMC2x. I don't mind to add CI test later on another PR but it should be for this release.

PS: @attie-argentum, @benys , can you help with some tests on real HW too?

photonthunder commented:
The easiest test is with mcuboot shell enabled (auto enabled with SHELL and MCUBOOT_IMAGE_MANAGER) and then type mcuboot request_upgrade.  Even with the cache controller enabled most of the testing I did with flash_writes were successful.  It would appear that the delay is greater and thus caught when it shows up at the end of the FLASH_BLOCK_SIZE.

photonthunder commented:
@nandojve, I have run full mcuboot swap update with image confirmation on reboot using #61005 with the changes in this PR and everything looks good.

nandojve commented:
Hi @photonthunder ,

I already tested on HW this changes and it passed. I would like you add attached patch as another commit to enable tests for sam0.

patch [add_sam0_flash_test_config.txt](https://github.com/zephyrproject-rtos/zephyr/files/12362116/add_sam0_flash_test_config.txt)

photonthunder committed:
boards: arm: atmel: Add flash support to SAM0 xpro boards

Add the following XPRO boards to flash driver test:
SAMD20, SAMDE54 and SAMR21

Signed-off-by: Daniel Evans <photonthunder@gmail.com>

Co-authored-by: Gerson Fernando Budke nandojve@gmail.com

photonthunder commented:
@nandojve, got it added.  There are some conflicts, since I am still on an older branch.  What is the proper Zephyr way to resolve them.  Do I add them to my commit or follow GitHub and do the merge commit?

photonthunder commented:
So I did a git reset --hard to the latest main for my branch and then just cherry-picked each commit and then forced push.

nandojve reviewed (changes requested):
Only minor cosmetics.
BTW, I would like you keep my credits on the last commit. Usually in this situations we add an entry as **_Co-authored-by: Gerson Fernando Budke <nandojve@gmail.com>_**

photonthunder commented:
@nandojve, fixed.  I was wondering how you all did that...  Kinda like the copyright thing on a file, I never really understand the rules of that either...

nandojve commented:
> @nandojve, fixed. I was wondering how you all did that... Kinda like the copyright thing on a file, I never really understand the rules of that either...

Hi @photonthunder ,

In this case you are the Author, credits are yours. Zephyr requires the Signed-off-by that tell about origin this code. You can read more about on [developer-certification-of-origin-dco](https://docs.zephyrproject.org/1.14.0/contribute/index.html#developer-certification-of-origin-dco). Since I shared code with you basically you tell that the code or part of the code is from and you credit the person/people that help you with that.

I could do it later but I think this is better and since we merge the changes and enable the tests.

nandojve commented:
> @nandojve, looks like the current issue is with spi_nor_configure when running the flash test. Not sure the best way to address that so I will await guidance...

I'll check it this week.
Thank you for pointed out.

nandojve commented:
Hi @photonthunder ,

rebase please!

nandojve commented:
Hi @photonthunder ,

You need fix compliance check.

photonthunder commented:
@nandojve, build errors that I think are related to CONFIG_MAX_THREAD_BYTES not getting assigned properly because of the lack of dependencies but the proper way to fix it I will leave to you...

nandojve commented:
Hi @photonthunder ,

Rebase on top of main to see if we get CI green.
